### ISSUE TITLE
JSONDecodeError

### ISSUE THREAD

**Issue by timlinden** [Taxel/PlexTraktSync#14]
On my first run I got the following error. It sync'd up until a certain movie and stopped. When I ran it again it continued on. Just letting you know ;-)

`Processing section Movies
Traceback (most recent call last):
  File "main.py", line 315, in <module>
    main()
  File "main.py", line 296, in main
    section, trakt_watched_movies, ratings, listutil, trakt_movie_collection)
  File "main.py", line 80, in process_movie_section
    m.add_to_library()
  File "C:\Users\timli\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\LocalCache\local-packages\Python37\site-packages\trakt\movies.py", line 261, in add_to_library
    add_to_collection(self)
  File "C:\Users\timli\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\LocalCache\local-packages\Python37\site-packages\trakt\core.py", line 507, in inner
    json_data = self._handle_request('post', url, data=args)
  File "C:\Users\timli\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\LocalCache\local-packages\Python37\site-packages\trakt\core.py", line 452, in _handle_request
    json_data = json.loads(response.content.decode('UTF-8', 'ignore'))
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.7_3.7.1776.0_x64__qbz5n2kfra8p0\lib\json\__init__.py", line 348, in loads
    return _default_decoder.decode(s)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.7_3.7.1776.0_x64__qbz5n2kfra8p0\lib\json\decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.7_3.7.1776.0_x64__qbz5n2kfra8p0\lib\json\decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)`

**Comment by gitthangbaby** [comment]
getting this nonstop. i now couldn't finish a sync after 200 attempts. also operation got extremelly slow, one hour per each section. the only chance is running it in a loop all day to get at least one library finished (with 'sync complete' status already). now i hope for more filtrers to come, it;s not commited yet.

**Comment by twolaw** [comment]
I get this error sometimes too.

**Comment by gitthangbaby** [comment]
i created catch blocks everywhere and repeat the action if failed, now i can always finish the sync, and in 170seconds instead of hours.

**Comment by twolaw** [comment]
Nice, I will try it.

Do you have http 502 errors from trakt too ?
I have 502 bad gateway answers sometimes, I don't know if it is linked.

**Comment by twolaw** [comment]
@gitthangbaby : I have run your main.py and it crashed.

Output:
```
error at try #1
error at try #2
error at try #3
error at try #4
Processing section Films
Traceback (most recent call last):
  File "./main.py", line 381, in <module>
    main()
  File "./main.py", line 361, in main
    section, trakt_watched_movies, ratings, listutil, trakt_movie_collection)
UnboundLocalError: local variable 'trakt_watched_movies' referenced before assignment
```
Logfile:

```
2020-01-09 14:24:06,227 INFO:Processing section Films
2020-01-09 14:24:09,915 INFO:requesting noncached data #1
2020-01-09 14:24:09,915 DEBUG:get: https://api-v2launch.trakt.tv/users/me
2020-01-09 14:24:09,916 DEBUG:headers: {'Content-Type': 'application/json', 'trakt-api-version': '2', 'trakt-api-key': None, 'Authorization': 'Bearer None'}
2020-01-09 14:24:09,917 DEBUG:method, url :: get, https://api-v2launch.trakt.tv/users/me
2020-01-09 14:24:09,926 DEBUG:Starting new HTTPS connection (1): api-v2launch.trakt.tv:443
2020-01-09 14:24:10,236 DEBUG:https://api-v2launch.trakt.tv:443 "GET /users/me HTTP/1.1" 403 None
2020-01-09 14:24:10,265 DEBUG:RESPONSE [get] (https://api-v2launch.trakt.tv/users/me): <Response [403]>
2020-01-09 14:24:10,266 ERROR:cannot read user ratings (1)
2020-01-09 14:24:10,267 INFO:requesting noncached data #2
2020-01-09 14:24:10,267 DEBUG:get: https://api-v2launch.trakt.tv/users/me
2020-01-09 14:24:10,268 DEBUG:headers: {'Content-Type': 'application/json', 'trakt-api-version': '2', 'trakt-api-key': None, 'Authorization': 'Bearer None'}
2020-01-09 14:24:10,268 DEBUG:method, url :: get, https://api-v2launch.trakt.tv/users/me
2020-01-09 14:24:10,277 DEBUG:Starting new HTTPS connection (1): api-v2launch.trakt.tv:443
2020-01-09 14:24:10,588 DEBUG:https://api-v2launch.trakt.tv:443 "GET /users/me HTTP/1.1" 403 None
2020-01-09 14:24:10,612 DEBUG:RESPONSE [get] (https://api-v2launch.trakt.tv/users/me): <Response [403]>
2020-01-09 14:24:10,613 ERROR:cannot read user ratings (2)
2020-01-09 14:24:10,614 INFO:requesting noncached data #3
2020-01-09 14:24:10,615 DEBUG:get: https://api-v2launch.trakt.tv/users/me
2020-01-09 14:24:10,616 DEBUG:headers: {'Content-Type': 'application/json', 'trakt-api-version': '2', 'trakt-api-key': None, 'Authorization': 'Bearer None'}
2020-01-09 14:24:10,617 DEBUG:method, url :: get, https://api-v2launch.trakt.tv/users/me
2020-01-09 14:24:10,630 DEBUG:Starting new HTTPS connection (1): api-v2launch.trakt.tv:443
2020-01-09 14:24:10,933 DEBUG:https://api-v2launch.trakt.tv:443 "GET /users/me HTTP/1.1" 403 None
2020-01-09 14:24:10,964 DEBUG:RESPONSE [get] (https://api-v2launch.trakt.tv/users/me): <Response [403]>
2020-01-09 14:24:10,964 ERROR:cannot read user ratings (3)
2020-01-09 14:24:10,965 INFO:requesting noncached data #4
2020-01-09 14:24:10,966 DEBUG:get: https://api-v2launch.trakt.tv/users/me
2020-01-09 14:24:10,966 DEBUG:headers: {'Content-Type': 'application/json', 'trakt-api-version': '2', 'trakt-api-key': None, 'Authorization': 'Bearer None'}
2020-01-09 14:24:10,966 DEBUG:method, url :: get, https://api-v2launch.trakt.tv/users/me
2020-01-09 14:24:10,974 DEBUG:Starting new HTTPS connection (1): api-v2launch.trakt.tv:443
2020-01-09 14:24:11,294 DEBUG:https://api-v2launch.trakt.tv:443 "GET /users/me HTTP/1.1" 403 None
2020-01-09 14:24:11,318 DEBUG:RESPONSE [get] (https://api-v2launch.trakt.tv/users/me): <Response [403]>
2020-01-09 14:24:11,319 ERROR:cannot read user ratings (4)
2020-01-09 14:24:11,320 ERROR:movies were not loaded
2020-01-09 14:24:11,321 DEBUG:Movie ratings from trakt: {}
```

I tried the original main.py again to check and it worked correctly this time. The trakt API sometimes don't answer and we should handle this in a better way.

**Comment by twolaw** [comment]
And why does it try to fetch ratings when I have set `ratings = false` in config ?

**Comment by gitthangbaby** [comment]
@twolaw i'm happy as my version now runs 100% without interruption. I also do scrubbing with the original plugin too. Thanks for this important script.
* the general problem is almost every call in this script is an online fetch (no idea why so prone to failing), so you get crashes in record number of lines. as above, trakt_movie_collection wasn't read, and next debug message collapses.. i mean.. no surprise next debug message was already commented before. oh wait, even this line was, i was too optimistic to uncomment it.  
* in your other case, you had too many retries. at one point, script should fail and this is the case. you can set your counters higher.
* ratings are read as before, you can see main() didn't change logic. and i couldn't spot an error like wrong indentation. i guess it doesn't really sync ratings just loads some info in main()...

**Comment by twolaw** [comment]
fixed with commit 71e1ebcf1e1011d51dbf508a0b1f85b093c0636e
